Public/Src/Engine/Scheduler/Graph/PipGraphFragmentManager.cs

```diff
@@ -0,0 +1,152 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics.ContractsLight;
+using System.Linq;
+using System.Threading.Tasks;
+using BuildXL.Pips;
+using BuildXL.Pips.Operations;
+using BuildXL.Scheduler.Tracing;
+using BuildXL.Utilities;
+using BuildXL.Utilities.Instrumentation.Common;
+
+namespace BuildXL.Scheduler.Graph
+{
+    /// <summary>
+    /// Manager which controls adding pip fragments to the graph.
+    /// </summary>
+    public class PipGraphFragmentManager : IPipGraphFragmentManager
+    {
+        private ConcurrentDictionary<string, (PipGraphFragmentSerializer, Task<bool>)> m_readFragmentTasks = new ConcurrentDictionary<string, (PipGraphFragmentSerializer, Task<bool>)>();
+
+        private IPipGraph m_pipGraph;
+
+        private PipExecutionContext m_context;
+
+        private PipGraphFragmentContext m_fragmentContext;
+
+        private LoggingContext m_loggingContext;
+
+        /// <summary>
+        /// PipGraphFragmentManager
+        /// </summary>
+        public PipGraphFragmentManager(LoggingContext loggingContext, PipExecutionContext context, IPipGraph pipGraph)
+        {
+            m_loggingContext = loggingContext;
+            m_context = context;
+            m_pipGraph = pipGraph;
+            m_fragmentContext = new PipGraphFragmentContext();
+        }
+
+        /// <summary>
+        /// Add a single pip graph fragment to the graph.
+        /// </summary>
+        /// <param name="fragmentName">Name of the fragment, for use with specifying dependencyNames</param>
+        /// <param name="filePath">Path of binary fragment file</param>
+        /// <param name="dependencyNames">Names of fragments it depends on</param>
+        /// <returns>Task that returns true if successfully added all pips in the fragment to the graph</returns>
+        public Task<bool> AddFragmentFileToGraph(string fragmentName, AbsolutePath filePath, string[] dependencyNames)

```

Comment:
>fragmentName [](start = 56, length = 12)

Please restrict the identifier of the fragments to valid Symbols

Reply:
Is there a good helper method to do this?

---
In reply to: [289564252](https://github.com/microsoft/BuildXL/pull/394#discussion_r289564252) [](ancestors = 289564252)

Reply:
`FullSymbol.TryCreate`

---
In reply to: [289574107](https://github.com/microsoft/BuildXL/pull/394#discussion_r289574107) [](ancestors = 289574107,289564252)

Reply:
ID is now generated, no longer needed.

---
In reply to: [289935992](https://github.com/microsoft/BuildXL/pull/394#discussion_r289935992) [](ancestors = 289935992,289574107,289564252)